src/selectors/entities/channel_categories.ts

```diff
@@ -189,6 +189,76 @@ export function makeFilterAutoclosedDMs(getAutocloseCutoff = getDefaultAutoclose
     );
 }
 
+export function makeFilterAutoclosedDMs(): (state: GlobalState, channels: Channel[], categoryType: string) => Channel[] {
+    return createSelector(
+        (state: GlobalState, channels: Channel[]) => channels,
+        (state: GlobalState, channels: Channel[], categoryType: string) => categoryType,
+        getMyPreferences,
+        shouldAutocloseDMs,
+        getCurrentChannelId,
+        (state: GlobalState) => state.entities.users.profiles,
+        getCurrentUserId,
+        getMyChannelMemberships,
+        (channels, categoryType, myPreferences, autocloseDMs, currentChannelId, profiles, currentUserId, myMembers) => {
+            if (categoryType !== CategoryTypes.DIRECT_MESSAGES) {
+                // Only autoclose DMs that haven't been assigned to a category
+                return channels;
+            }
+
+            let unreadCount = 0;
+
+            const filtered = channels.filter((channel) => {
+                if (channel.type !== General.DM_CHANNEL && channel.type !== General.GM_CHANNEL) {
+                    return true;
+                }
+
+                if (isUnreadChannel(myMembers, channel)) {
+                    unreadCount++;
+
+                    // Unread DMs/GMs are always visible
+                    return true;
+                }
+
+                if (currentChannelId === channel.id) {
+                    // The current channel is always visible
+                    return true;
+                }
+
+                // openTime is the time the channel was last opened (like from the More DMs list) after having been closed
+                const openTimePref = myPreferences[getPreferenceKey(Preferences.CATEGORY_CHANNEL_OPEN_TIME, channel.id)];
+                const openTime = parseInt(openTimePref ? openTimePref.value! : '0', 10);
+
+                // DMs with deactivated users will be visible if you're currently viewing them and they were opened
+                // since the user was deactivated
+                if (channel.type === General.DM_CHANNEL) {
+                    const teammateId = getUserIdFromChannelName(currentUserId, channel.name);
+                    const teammate = profiles[teammateId];
+
+                    if (!teammate || teammate.delete_at > openTime) {
+                        return false;
+                    }
+                }
+
+                return true;
+            });
+
+            // The limit of DMs user specifies to be rendered in the sidebar
+            const limitPref = myPreferences[getPreferenceKey(Preferences.CATEGORY_SIDEBAR_SETTINGS, Preferences.LIMIT_VISIBLE_DMS_GMS)];
+            const limit = parseInt(limitPref ? limitPref.value! : '0', 10);
+
+            let remaining;
+
+            if (limit < unreadCount) {
+                remaining = unreadCount;
+            } else {
+                remaining = limit;
+            }
+
+            return filtered.slice(0, remaining);
```

Comment:
This isn't going to work unfortunately since it'll slice the first X channels alphabetically or with the X most recent posts (based on category sorting) instead of the last X viewed. We'll need to do something like constructing an array of the last X viewed channels and then use that as part of the above filter instead.

Reply:
Also, if you could add the `return filtered.length === channels.length ? channels : filtered;` check like above, that'd help reduce re-renders if this doesn't actually end up filtering anything out.